File: CSE101/pa1/Lex.c
//-----------------------------------------------------------------------------
// Lex.c
// Zachary Dutra
// zdutra
// 1581789
// pa1
//-----------------------------------------------------------------------------

#include<stdio.h>
#include<stdlib.h>
#include<string.h>
#include "List.h"

#define MAX_LEN 5000

int main(int argc, char * argv[]){

   int count = 0;
   int compare = 0;
   FILE *in, *out;
   char line[MAX_LEN];
   char* token;
   char** in_list = (char**) calloc(MAX_LEN, sizeof(char));
   // int compare, count = 0;
   int n = 1;
   int linecount = 0;

   // check command line for correct number of arguments
   if( argc != 3 ){
      printf("Usage: %s <input file> <output file>\n", argv[0]);
      exit(1);
   }

   // open files for reading and writing 
   in = fopen(argv[1], "r");
   out = fopen(argv[2], "w");
   if( in==NULL ){
      printf("Unable to open file %s for reading\n", argv[1]);
      exit(1);
   }
   if( out==NULL ){
      printf("Unable to open file %s for writing\n", argv[2]);
      exit(1);
   }

   /* read each line of input file, then count and print tokens */
   for(int j = 0; fgets(line, MAX_LEN, in) != NULL; j++)  {
      linecount++;
      token = strtok(line, "\n");
      int length = strlen(token);
      in_list[j] = calloc(length, sizeof(char*));
      strcpy(in_list[j], token);
   }
   List Lex_list = newList();
   prepend(Lex_list, 0);
   moveFront(Lex_list);
   while(n < linecount){
      compare = strcmp(in_list[n], in_list[get(Lex_list)]);
      if(compare <= 0){
         insertBefore(Lex_list, n);
         n++;
         count++;
         moveFront(Lex_list);
         continue;
      }
      else if(compare > 0 && get(Lex_list) != back(Lex_list)){
         moveNext(Lex_list);
         continue;
      }
      else{
         append(Lex_list, n);
         n++;
         count++;
         moveFront(Lex_list);
      }
   }
   for(int i = 0; i < n; i++){
      fprintf(out,"%s\n", in_list[get(Lex_list)]);
      moveNext(Lex_list);
   }
   /* close files */
   fclose(in);
   fclose(out);

   return(0);
}

File: pa1_Java/Lex.java
/*
Zachary Dutra
zdutra
1581789
pa1
*/

import java.util.*;
import java.io.*;

class Lex{
    public static void main(String[]args)throws IOException{
        PrintWriter out;
        Scanner in;
        String line;
        String[] token;
        int n = 0;
        int comp = 0;
        int lineNumber = 0;
        int j = 1;
        int count = 1;
        int stringcount = 0;
        if(args.length < 2){
            System.err.println("Usage: FileIO infile outfile");
            System.exit(1);
        }
        FileReader fr = new FileReader(args[0]);
        LineNumberReader lnr = new LineNumberReader(fr);
        in = new Scanner(new File(args[0]));
        FileWriter output = new FileWriter(new File(args[1]));
        out = new PrintWriter(output);
        while(lnr.readLine()!=null){
            lineNumber++;
        }
        String[] a_list = new String[lineNumber];
        while(in.hasNextLine()){
            line = in.nextLine();
            a_list[stringcount] = line;
            stringcount++;
        }
            List LexList = new List();
            LexList.prepend(0);
            LexList.moveFront();
            while(count < lineNumber){
                comp = a_list[j].compareTo(a_list[LexList.cursor.data]);
                if(comp <= 0){
                    LexList.insertBefore(j);
                    j++;
                    count++;
                    LexList.moveFront();
                    continue;
                }
                else if(comp>0 && LexList.cursor != LexList.back){
                    LexList.moveNext();
                    continue;
                }
                else{
                    LexList.append(j);
                    j++;
                    count++;
                    LexList.moveFront();
                }
            }
        Node printNode = LexList.front;
        for(int i = 0; i < a_list.length; i++){
            out.println(a_list[printNode.data]);
            printNode = printNode.next;
        }
        in.close();
        output.close();
        out.close();
    }
}
